cmd/secretcli/secret20.go

```diff
@@ -0,0 +1,371 @@
+
+	cmd := &cobra.Command{
+		Use:   "deposit [contract address or label]",
+		Short: "Convert your SCRT into a secret token",
+		Long:  `Convert your SCRT into a secret token. This command will only work if the token supports native currency conversion`,
+		Args:  cobra.MinimumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			inBuf := bufio.NewReader(cmd.InOrStdin())
+			cliCtx := context.NewCLIContextWithInput(inBuf).WithCodec(cdc)
+
+			contractAddrStr, err := addressFromBechOrLabel(args[0], cliCtx)
+			if err != nil {
+				return err
+			}
+
+			contractAddr, err := sdk.AccAddressFromBech32(contractAddrStr)
+			if err != nil {
+				return errors.New("invalid contract address or label")
+			}
+
+			msg := depositMsg()
+
+			amountStr := viper.GetString(flagAmount)
+
+			return cli.ExecuteWithData(cmd, contractAddr, msg, amountStr, false, "", "", cdc)
+		},
+	}
+	cmd.Flags().String(flagAmount, "", "")
+	return cmd
+}
+
+func s20Withdraw(cdc *codec.Codec) *cobra.Command {
+
+	cmd := &cobra.Command{
+		Use:   "withdraw [contract address or label] [amount]",
+		Short: "Convert your secret token back to SCRT",
+		Long:  `Convert your secret token back to SCRT. This command will only work if the token supports native currency conversion`,
+		Args:  cobra.MinimumNArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			inBuf := bufio.NewReader(cmd.InOrStdin())
+			cliCtx := context.NewCLIContextWithInput(inBuf).WithCodec(cdc)
+
+			contractAddrStr, err := addressFromBechOrLabel(args[0], cliCtx)
+			if err != nil {
+				return err
+			}
+
+			contractAddr, err := sdk.AccAddressFromBech32(contractAddrStr)
+			if err != nil {
+				return errors.New("invalid contract address or label")
+			}
+
+			amount := args[1]
+
+			msg := withdrawMsg(amount)
+
+			return cli.ExecuteWithData(cmd, contractAddr, msg, "", false, "", "", cdc)
+		},
+	}
+
+	return cmd
+}
+
+func transferHistoryMsg(fromAddress sdk.AccAddress, viewingKey string) []byte {
+	return []byte(fmt.Sprintf("{\"transfers\": {\"address\": \"%s\", \"key\": \"%s\"}}", fromAddress.String(), viewingKey))
+}
+
+func balanceMsg(fromAddress sdk.AccAddress, viewingKey string) []byte {
+	return []byte(fmt.Sprintf("{\"balance\": {\"address\": \"%s\", \"key\": \"%s\"}}", fromAddress.String(), viewingKey))
+}
+
+func sendCoinMsg(toAddress sdk.AccAddress, amount string) []byte {
+	return []byte(fmt.Sprintf("{\"transfer\": {\"recipient\": \"%s\", \"amount\": \"%s\"}}", toAddress.String(), amount))
+}
+
+func depositMsg() []byte {
+	return []byte(fmt.Sprintf("{\"deposit\": {}}"))
+}
+
+func withdrawMsg(amount string) []byte {
+	return []byte(fmt.Sprintf("{\"withdraw\": {\"amount\": \"%s\"}}", amount))
+}
+
+func createViewingKeyMsg(data string) []byte {
+	return []byte(fmt.Sprintf("{\"create_viewing_key\": {\"entropy\": \"%s\"}}", data))
+}
+
+func setViewingKeyMsg(data string) []byte {
+	return []byte(fmt.Sprintf("{\"set_viewing_key\": {\"key\": \"%s\"}}", data))
+}
```

Comment:
It's probably a good idea to use proper json escaping here, using a json library, but bad inputs will just cause the contract to fail, so it's probably ok?

Reply:
The checks you added that it is a valid json will probably make this fail gracefully anyway

Reply:
That's fair.
I still find it ugly, but it works so I'm not gonna pursue this right now.